src/main/java/com/google/devtools/build/lib/analysis/actions/StrippingPathMapper.java

```diff
@@ -0,0 +1,262 @@
+ * qualifying actions before running them, thus improving caching. "Qualifying" actions are actions
+ * known not to depend on the names of their input and output paths. Non-qualifying actions include
+ * manifest generators and compilers that store debug symbol source paths.
+ *
+ * <p>As an experimental feature, most logic is centralized here to provide easy hooks into executor
+ * and action code and avoid complicating large swaths of the code base.
+ *
+ * <p>Enable this feature by setting {@code --experimental_output_paths=strip}. This activates two
+ * effects:
+ *
+ * <ol>
+ *   <li>"Qualifying" actions strip config paths from their command lines. An action qualifies if
+ *       its implementation logic uses {@link PathMappers#create(Action, OutputPathsMode)} as
+ *       described in its javadocs and has its mnemonic listed in {@link
+ *       PathMappers#SUPPORTED_MNEMONICS}. Such an action must pass the {@link PathMapper} to all
+ *       structured command line constructions. If any unstructured command line arguments refer to
+ *       artifact paths, custom handling needs to be added to {@code mapCustomStarlarkArgv} or
+ *       {@code getMapFn} below.
+ *   <li>A supporting executor strips paths from qualifying actions' inputs and outputs before
+ *       staging for execution by taking {@link Spawn#getPathMapper()} into account.
+ * </ol>
+ *
+ * <p>So an action is responsible for declaring that it strips paths and adjusting its command line
+ * accordingly. The executor is responsible for remapping action inputs and outputs to match.
+ */
+public final class StrippingPathMapper {
+  static final String GUID = "8eb2ad5a-85d4-435b-858f-5c192e91997d";
+
+  /**
+   * Creates a new {@link PathMapper} that strips config prefixes if the particular action instance
+   * supports it.
+   *
+   * @param action the action to potentially strip paths from
+   * @return a {@link StrippingPathMapper} if the action supports it, else {@link Optional#empty()}.
+   */
+  static Optional<PathMapper> tryCreate(Action action) {
+    // This is expected to always be "bazel-out", but we don't want to hardcode it here.
+    PathFragment outputRoot = action.getPrimaryOutput().getExecPath().subFragment(0, 1);
+    if (isPathStrippable(action.getInputs(), outputRoot)) {
+      return Optional.of(
+          create(action.getMnemonic(), action instanceof StarlarkAction, outputRoot));
+    }
+    return Optional.empty();
+  }
+
+  private static PathMapper create(String mnemonic, boolean isStarlarkAction,
+      PathFragment outputRoot) {
+    final StringStripper argStripper = new StringStripper(outputRoot.getPathString());
+    final MapFn<Object> structuredArgStripper = (object, args) -> {
+      if (object instanceof String) {
+        args.accept(argStripper.strip((String) object));
+      } else {
+        args.accept(CommandLineItem.expandToCommandLine(object));
+      }
+    };
+    final boolean isJavaAction =
```

Comment:
This is only for the purpose of handling `--javacopts`, right?

I realize there are some specific action references in this file. Over time I think we want to aim for eliminating such references in preference for embedding them in the action definition itself.